src/types/tuple.rs

```diff
@@ -276,23 +509,103 @@ impl<'a> DoubleEndedIterator for PyTupleIterator<'a> {
     }
 }
 
-impl<'a> ExactSizeIterator for PyTupleIterator<'a> {
+impl<'py> ExactSizeIterator for PyTupleIterator2<'py> {
     fn len(&self) -> usize {
         self.length.saturating_sub(self.index)
     }
 }
 
-impl FusedIterator for PyTupleIterator<'_> {}
+impl FusedIterator for PyTupleIterator2<'_> {}
 
-impl<'a> IntoIterator for &'a PyTuple {
-    type Item = &'a PyAny;
-    type IntoIter = PyTupleIterator<'a>;
+impl<'py> IntoIterator for &'_ Py2<'py, PyTuple> {
+    type Item = Py2<'py, PyAny>;
+    type IntoIter = PyTupleIterator2<'py>;
 
     fn into_iter(self) -> Self::IntoIter {
         self.iter()
     }
 }
 
+impl<'py> IntoIterator for Py2<'py, PyTuple> {
+    type Item = Py2<'py, PyAny>;
+    type IntoIter = PyTupleIterator2<'py>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        PyTupleIterator2::new(self)
+    }
+}
+
+/// Used by `PyTuple::iter_borrowed()`.
+pub(crate) struct PyTupleIterBorrowed<'a, 'py> {
+    tuple: Py2Borrowed<'a, 'py, PyTuple>,
+    index: usize,
+    length: usize,
+}
+
+impl<'a, 'py> PyTupleIterBorrowed<'a, 'py> {
+    fn new(tuple: Py2Borrowed<'a, 'py, PyTuple>) -> Self {
+        let length = tuple.len();
+        PyTupleIterBorrowed {
+            tuple,
+            index: 0,
+            length,
+        }
+    }
+
+    unsafe fn get_item(
+        tuple: Py2Borrowed<'a, 'py, PyTuple>,
+        index: usize,
+    ) -> Py2Borrowed<'a, 'py, PyAny> {
+        #[cfg(any(Py_LIMITED_API, PyPy))]
+        let item = tuple.get_borrowed_item(index).expect("tuple.get failed");
+        #[cfg(not(any(Py_LIMITED_API, PyPy)))]
+        let item = tuple.get_borrowed_item_unchecked(index);
+        item
+    }
+}
+
+impl<'a, 'py> Iterator for PyTupleIterBorrowed<'a, 'py> {
+    type Item = Py2Borrowed<'a, 'py, PyAny>;
+
+    #[inline]
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.index < self.length {
+            let item = unsafe { Self::get_item(self.tuple, self.index) };
```

Comment:
This is where the `Py2Borrowed` copying occurs, but I can rework this if adding copying now seems too soon.

Reply:
I certainly agree that `Borrowed` should `Copy` like shared references which I think we already merged in an earlier PR.

Reply:
Yep we merged this already in #3680 as the `Copy` property kept showing up as useful.